kong/developer_service.go

```diff
@@ -0,0 +1,197 @@
+package kong
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+)
+
+// AbstractDeveloperService handles Developers in Kong.
+type AbstractDeveloperService interface {
+	// Create creates a Developer in Kong.
+	Create(ctx context.Context, developer *Developer) (*Developer, error)
+	// Get fetches a Developer in Kong.
+	Get(ctx context.Context, emailOrID *string) (*Developer, error)
+	// GetByCustomID fetches a Developer in Kong.
+	GetByCustomID(ctx context.Context, customID *string) (*Developer, error)
+	// Update updates a Developer in Kong
+	Update(ctx context.Context, developer *Developer) (*Developer, error)
+	// Delete deletes a Developer in Kong
+	Delete(ctx context.Context, emailOrID *string) error
+	// List fetches a list of Developers in Kong.
+	List(ctx context.Context, opt *ListOpt) ([]*Developer, *ListOpt, error)
+	// ListAll fetches all Developers in Kong.
+	ListAll(ctx context.Context) ([]*Developer, error)
+}
+
+// DeveloperService handles Developers in Kong.
+type DeveloperService service
+
+// Create creates a Developer in Kong.
+// If an ID is specified, it will be used to
+// create a developer in Kong, otherwise an ID
+// is auto-generated.
+func (s *DeveloperService) Create(ctx context.Context,
+	developer *Developer) (*Developer, error) {
+
+	queryPath := "/developers"
+	method := "POST"
```

Comment:
In case of basic_auth you'll need a password field to create a consumer otherwise, you'll get a bad request. What I would recommand is to generate a random password value  (uuid for example)  just for the creation of the developer in case of basic-auth.
In openid-connect, the password is not required

Reply:
The solution might simply be to add the password field to the Developer structure

Reply:
This presents a larger issue. The way developer management is implemented isn't particular conducive to directly manipulating them via the API, as their API calls can include fields that manipulate other objects.

The password, for example, isn't a developer field: while you send it in a `/developers` POST, it actually ends up in a basic-auth credential, but not one that's visible (you can't see it on the developer entity and the consumer it's attached to is hidden).

go-kong can probably gloss over that, since it can just send a request with a password field anyway, and will just omit it when GETs don't return one. If you were to try to then use this in decK, however, it'd try to sync the developer on every pass, since the password would be missing as far as it's concerned.